docs/WorkerAgent.md

```diff
@@ -0,0 +1,44 @@
+
+The Agent is responsible for the following:
+1. Handle GRPC requests to perform work from Server.
+2. Spawn a process with the appropriate Lane name, configured with the correct execution user, group, and environment.
+3. Connect the output of the command to GRPC response to the Server.
+4. Report back the exit status of the child process.
+5. Deliver artifacts to any external persistence services.
+
+## Process Environment
+Each GRPC request to the Agent should contain all of the necessary configuration information required to run the Lane successfully. This can be accomplished by supplying an environment (the ENV variables) to the child process when it is spawned by the Agent.
+
+<img src="WorkerAgent.svg">
+
+For more on using the environment for configuration (https://12factor.net/config)
+The Agent may have to fetch resources (such as source code) from other services but it should not have to fetch any configuration data.
+
+## Permission model
+The Agent will be required to run arbitrary Ruby code on the Worker. It may also be required to perform operations that require root such as installing Xcode. While there will be other security mechanisms that will enforce strict security against malicious users, we can help ourselves by preventing accidental (or incidental) damage to the Worker by running the Agents under a special user that belongs to the sudoers group. This will allow us to whitelist the commands and permissions the user can perform on a Worker.
+
+## Throttling
+Each agent is capable of spawning as many processes to handle the requests as it wants, but there are stability considerations that we should take into account. Lane execution  may be very resource heavy and launching too many of them may make the Worker kill the process prematurely. There are two possibilities for implementing throttling:
+1. Implement a serial queue on the Agent
+2. The Agent reports its capacity to the Server and rejects requests that cannot be fulfilled.
+
+Given that there will probably be a queueing service in the Server that will be responsible for distributing workload across many workers, it seems like the Agent should not implement another queue. The logic for job queueing should live somewhere higher in the stack and the Agent should simply be able to report its capacity when queried.
+
+## Persistence
+Workers should be ephemeral and all artifacts need to be placed somewhere with strong persistence during the execution. Logs should be streamed as part of the GRPC response. After a successful fastlane run, other artifacts such as IPA files and screenshots should be persisted outside of the Worker. Additionally the raw log file should be persisted as well, to access them in the future.
```

Comment:
💯 
Any thoughts on the best way to do this? Should we stick with GCS for now? Maybe we pass in a bucket, api key, etc along with a path?

Reply:
Depending on which `ArtifactProvider` is the user using. So my thoughts here is to provide the ability for those `ArtifactProviders` is to store the artifacts in the parent process by adding a functionality to send the artifact to the parent and him storing it, or provide the ability to each worker with all the information needed to store the artifacts by themselves.

Reply:
I think there is a high-level open question here. In my mind, there are 2 options.

1.) A lane that creates an artifact puts it somewhere in a temporary place that the Agent knows about. It is the job of the Agent to deliver the artifact in a permanent place. This would be something configured in the webapp.
2.) The lane defines where the artifacts go, possibly using some new fastlane actions. This would be configured in the Fastfile.

I think we should revisit this in the near feature when we start scoping out artifacts as a whole.

Reply:
Cool, I actually don't think fastlane should care at all (or lanes). It's totally a CI thing, so it's more of a CI general setting that a project can override. 

I think it might make sense for each build runner to do the storage

Reply:
So for the `LocalArtifactProvider` and others, we have to add a functionality that allows the Agent to send the arftifacts to the main server. IMO, I think the main node should gather all the artifacts from the Agents, and store them (as the main node knows which `ArtifactProvider` and its configuration is)